### PR TITLE
Utility to remove duplicate arguments for calls and callees

### PR DESCRIPTION
**Pull request by MichaelSt98**
Transform/convert e.g., 

```fortran
 SUBROUTINE driver (nlon, nlev, nb, var)
  USE kernel_mod, ONLY: kernel
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: nlon, nlev, nb
  REAL, INTENT(INOUT) :: var(nlon, nlev, 5, nb)
  INTEGER :: ibl
  INTEGER :: offset
  INTEGER :: some_val
  INTEGER :: loop_start, loop_end
  loop_start = 2
  loop_end = nb
  some_val = 0
  offset = 1
!$omp test
  DO ibl=loop_start,loop_end
    CALL kernel(nlon, nlev, var(:, :, 1, ibl), var(:, :, 1, ibl), var(:, :, 2:5, ibl), offset, loop_start, loop_end, nlev)
    CALL kernel(nlon, nlev, var(:, :, 1, ibl), var(:, :, 1, ibl), var(:, :, 2:5, ibl), offset, loop_start, loop_end, nlev)
  END DO
END SUBROUTINE driver
SUBROUTINE kernel (nlon, nlev, var1, var2, another_var, icend, lstart, lend, kend)
  USE compute_mod, ONLY: compute
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: nlon, nlev, icend, lstart, lend, kend
  REAL, INTENT(INOUT) :: var1(nlon, nlev)
  REAL, INTENT(INOUT) :: var2(nlon, nlev)
  REAL, INTENT(INOUT) :: another_var(nlon, nlev, 4)
  INTEGER :: jk, jl, jt
  var1(:, :) = 0.
  DO jk=1,kend
    DO jl=1,nlon
      var1(jl, jk) = 0.
      var2(jl, jk) = 1.0
      DO jt=1,4
        another_var(jl, jk, jt) = 0.0
      END DO
    END DO
  END DO
  CALL compute(nlon, nlev, var1, var2)
  CALL compute(nlon, nlev, var1, var2)
END SUBROUTINE kernel
SUBROUTINE compute (nlon, nlev, b_var, a_var)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: nlon, nlev
  REAL, INTENT(INOUT) :: b_var(nlon, nlev)
  REAL, INTENT(INOUT) :: a_var(nlon, nlev)
  b_var(:, :) = 0.
  a_var(:, :) = 1.0
END SUBROUTINE compute
```

to

```fortran
SUBROUTINE driver (nlon, nlev, nb, var)
  USE kernel_mod, ONLY: kernel
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: nlon, nlev, nb
  REAL, INTENT(INOUT) :: var(nlon, nlev, 5, nb)
  INTEGER :: ibl
  INTEGER :: offset
  INTEGER :: some_val
  INTEGER :: loop_start, loop_end
  loop_start = 2
  loop_end = nb
  some_val = 0
  offset = 1
!$omp test
  DO ibl=loop_start,loop_end
    CALL kernel(nlon, nlev, var(:, :, 1, ibl), var(:, :, 2:5, ibl), offset, loop_start, loop_end)
    CALL kernel(nlon, nlev, var(:, :, 1, ibl), var(:, :, 2:5, ibl), offset, loop_start, loop_end)
  END DO
END SUBROUTINE driver
SUBROUTINE kernel (nlon, nlev, var, another_var, icend, lstart, lend)
  USE compute_mod, ONLY: compute
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: nlon, nlev, icend, lstart, lend
  REAL, INTENT(INOUT) :: var(nlon, nlev)
  REAL, INTENT(INOUT) :: another_var(nlon, nlev, 4)
  INTEGER :: jk, jl, jt
  var(:, :) = 0.
  DO jk=1,nlev
    DO jl=1,nlon
      var(jl, jk) = 0.
      var(jl, jk) = 1.0
      DO jt=1,4
        another_var(jl, jk, jt) = 0.0
      END DO
    END DO
  END DO
  CALL compute(nlon, nlev, var)
  CALL compute(nlon, nlev, var)
END SUBROUTINE kernel
SUBROUTINE compute (nlon, nlev, var)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: nlon, nlev
  REAL, INTENT(INOUT) :: var(nlon, nlev)
  var(:, :) = 0.
  var(:, :) = 1.0
END SUBROUTINE compute
```

### CODE REVIEW

File: loki/transformations/utilities.py [hunk 1]
```diff
@@ -32,10 +35,167 @@
...
+        this won't work properly for multiple calls to the same routine
+        with differing duplicate arguments
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine to be transformed.
+    rename_common : bool, optional
+        Try to rename duplicate arguments by finding a common pattern
+        in those names (Default: `False`).
```
Comment (reuterbal) [inline loki/transformations/utilities.py:89]: This relates to the callee side only, right?

It took me a while to figure out what the effect of this is, maybe we can expand the description like this?
```suggestion
        Try to rename dummy arguments in called routines that received the same argument
        on the caller side, by finding a common name pattern in those names (Default: `False`).
```

File: loki/transformations/utilities.py [hunk 2]
```diff
@@ -32,10 +35,167 @@
...
+    and callee. 
+
+    .. warning::
+        this won't work properly for multiple calls to the same routine
+        with differing duplicate arguments
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine to be transformed.
```
Comment (reuterbal) [inline loki/transformations/utilities.py:86]:
```suggestion
        The subroutine where calls should be transformed.
```

File: loki/transformations/utilities.py [hunk 3]
```diff
@@ -32,10 +35,167 @@
...
+        # (filter duplicate arguments and) update call
+        call._update(arguments=as_tuple(dict.fromkeys(call.arguments)), kwarguments=new_kwargs)
+        return arg_map
+
+    def modify_callee(callee, callee_arg_map):
+        combine = [routine_args for call_arg, routine_args in callee_arg_map.items() if len(routine_args) > 1]
+        matches = [os.path.commonprefix([str(elem.name) for elem in args]).rstrip('_')
+                or os.path.commonprefix([str(elem.name)[::-1] for elem in args]).rstrip('_')[::-1] for args in combine]
+        rename_common_map = {combine[i][0].name: matches[i] for i in range(len(combine)) if matches[i] != ''}\
+                if rename_common else {}
```
Comment (reuterbal) [inline loki/transformations/utilities.py:109]: For readability, I would prefer not to do use an inline `if` here. Wrapping this in proper conditional branches also avoids building `matches` if it's not required.

Secondly, the common pattern for `combine[i]: matches[i] for i in range(len(combine))` is to use `zip`:
```suggestion
        if rename_common:
            matches = [
                os.path.commonprefix([str(elem.name) for elem in args]).rstrip('_') or
                os.path.commonprefix([str(elem.name)[::-1] for elem in args]).rstrip('_')[::-1]
                for args in combine
            ]
            rename_common_map = {c[0].name: m for c, m in zip(combine, matches) if m}
        else:
            rename_common_map = {}
```
```

File: loki/transformations/utilities.py [hunk 4]
```diff
@@ -32,10 +35,167 @@
...
+        _new_kwargs = as_tuple(list(_)[0] for g, _ in it.groupby(call.kwarguments, key=lambda x: x[1]))
+        # filter duplicate kwargs (comparing to the arguments)
+        new_kwargs = tuple(kwarg for kwarg in _new_kwargs if kwarg[1] not in call.arguments)
+        # (filter duplicate arguments and) update call
+        call._update(arguments=as_tuple(dict.fromkeys(call.arguments)), kwarguments=new_kwargs)
+        return arg_map
+
+    def modify_callee(callee, callee_arg_map):
+        combine = [routine_args for call_arg, routine_args in callee_arg_map.items() if len(routine_args) > 1]
+        matches = [os.path.commonprefix([str(elem.name) for elem in args]).rstrip('_')
```
Comment (reuterbal) [inline loki/transformations/utilities.py:106]: Also, I'm wondering what happens if the "guessed" common name has a clash with an existing dummy argument. 
For example, what if `kernel` in the test had another dummy arg `var` that does not receive the common parameter. For example, `call kernel(var=1.0, var1=var(...), var2=var(...), ...)`.

File: loki/transformations/utilities.py [hunk 5]
```diff
@@ -32,10 +35,167 @@
...
+    rename_common_map_routine = {}
+    # modify/adapt callees
+    for callee, callee_arg_map in call_arg_map.items():
+        rename_common_map_routine[str(callee.name).lower()] = modify_callee(callee, callee_arg_map)
+    # handle possibly renamed kwarguments on caller side
+    if rename_common:
+        rename_kwarguments(relevant_calls, rename_common_map_routine)
+
+
+def modify_variable_declarations(routine, remove_symbols=(), rename_symbols=None):
```
Comment (reuterbal) [inline loki/transformations/utilities.py:162]: This partially re-implements the setter for `ProgramUnit,.variables`: https://github.com/ecmwf-ifs/loki/blob/cc179b3b86613af86df6dabb0be250cde4414fc9/loki/program_unit.py#L505

Should we maybe make this functionality a method on `ProgramUnit` and use it in both cases? (Cc @mlange05 for input)

**Review by reuterbal — CHANGES_REQUESTED** [review]
First of all, apologies for the appallingly long time it took me to review this!

This is a really nice and useful utility and a prime example of the inter-procedural way of working in Loki. I think the implementation could be improved in a few places to make it more pythonic and I've left comments in the code.

From a functionality point of view, I left a question whether this sufficiently safeguards against accidental name clashes when enabling the `rename_common` mode.

File: loki/transformations/utilities.py (hunk 4 above)
Comment (MichaelSt98) [reply loki/transformations/utilities.py:106]: I'll implement a solution for that 👍

**Comment by MichaelSt98** [comment]
@reuterbal all your comments and requested changes should be addressed except for the `modify_variable_declarations` re-implementing partially the setter in `ProgramUnit.variables`

**Review by reuterbal — APPROVED** [review]
Many thanks for implementing the suggested changes. I think this looks ready now!

**Pull request merged**